### PR TITLE
GPU: Ignore color write mask if unusable

### PR DESCRIPTION
**Pull request by unknownbrackets**
We unconditionally attempt to use a shader with color write mask, so we can't just refuse to generate a shader - it'll mean non-supporting devices get no shader at all.

It's better to continue ignoring the mask in that case.

-[Unknown]

### CODE REVIEW

File: GPU/Common/FragmentShaderGenerator.cpp [hunk 1]
```diff
@@ -81,12 +81,7 @@ bool GenerateFragmentShader(const FShaderID &id, char *buffer, const ShaderLangu
 	bool doFlatShading = id.Bit(FS_BIT_FLATSHADE) && !bugs.Has(Draw::Bugs::BROKEN_FLAT_IN_SHADER);
 	bool shaderDepal = id.Bit(FS_BIT_SHADER_DEPAL);
 	bool bgraTexture = id.Bit(FS_BIT_BGRA_TEXTURE);
-	bool colorWriteMask = id.Bit(FS_BIT_COLOR_WRITEMASK);
-
-	if (colorWriteMask && !compat.bitwiseOps) {
-		*errorString = "Color Write Mask requires bitwise ops";
-		return false;
-	}
+	bool colorWriteMask = id.Bit(FS_BIT_COLOR_WRITEMASK) && compat.bitwiseOps;
```
Comment (hrydgard) [inline GPU/Common/FragmentShaderGenerator.cpp:84]: I intended this to be checked during ID generation so the failure case should not happen, but apparently didn't implement that huh, good catch. Not sure where it makes the most sense to do the check, but this is fine.

File: GPU/Common/FragmentShaderGenerator.cpp [hunk 2]
```diff
@@ -819,10 +814,9 @@ bool GenerateFragmentShader(const FShaderID &id, char *buffer, const ShaderLangu
 						WRITE(p, "  vec3 colortest = roundAndScaleTo255v(v.rgb);\n");
 						WRITE(p, "  if ((colortest.r %s u_alphacolorref.r) && (colortest.g %s u_alphacolorref.g) && (colortest.b %s u_alphacolorref.b)) %s\n", test, test, test, discardStatement);
 					} else if (compat.bitwiseOps) {
-						// Apparently GLES3 does not support vector bitwise ops.
 						WRITE(p, "  ivec3 v_scaled = roundAndScaleTo255iv(v.rgb);\n");
 						if (compat.shaderLanguage == GLSL_VULKAN) {
-							// TODO: Use this for GL as well?
+							// Apparently GLES3 does not support vector bitwise ops, but Vulkan does?
```
Comment (hrydgard) [inline GPU/Common/FragmentShaderGenerator.cpp:819]: Oh, right, it's not just about bitwise but vector bitwise which was weirdly left out of ES3?

**Review by hrydgard — APPROVED** [review]

File: GPU/Common/FragmentShaderGenerator.cpp (hunk 2 above)
Comment (unknownbrackets) [reply GPU/Common/FragmentShaderGenerator.cpp:819]: Correct, ivec3 & ivec3 is not supported, btu int & int is.  So we have to do it by component outside Vulkan.

-[Unknown]

File: GPU/Common/FragmentShaderGenerator.cpp [hunk 3]
```diff
@@ -422,13 +422,14 @@ bool GenerateFragmentShader(const FShaderID &id, char *buffer, const ShaderLangu
 	if (colorWriteMask && !hasPackUnorm4x8) {
 		WRITE(p, "uint packUnorm4x8(vec4 v) {\n");
 		WRITE(p, "  v = clamp(v, 0.0, 1.0);\n");
-		WRITE(p, "  uvec4 u = uvec4(255.0 * v);\n");
+		WRITE(p, "  uvec4 u = uvec4(255 * uint(v));\n");
```
Comment (hrydgard) [inline GPU/Common/FragmentShaderGenerator.cpp:425]: This is not going to work. uint(v) kills the fractional part so it'll slam it to 0 or 255.

Comment (unknownbrackets) [reply GPU/Common/FragmentShaderGenerator.cpp]: Oops, forgot to finish that change.

-[Unknown]

Comment (unknownbrackets) [reply GPU/Common/FragmentShaderGenerator.cpp]: I had been thinking to convert it to a uint first at some point, but indeed it needs to be converted first.  Then I was thinking about the highp cast and got distracted by the other func, heh.

-[Unknown]

Comment (hrydgard) [reply GPU/Common/FragmentShaderGenerator.cpp]: Hehe, it's quite intricate stuff :)  I made many mistakes too before I got those Outrun reflections working...

**Pull request merged**